### PR TITLE
Raffle

### LINKED ISSUE #54: Add RaffleEntity

### PR DESCRIPTION
**Pull request by msvrtan**
In order to motivate members to leave feedback to our speakers
For meetup organizers ,
We will implement raffling logic

 ## References
 
Closes #54

### CODE REVIEW

File: features/picking-events-to-raffle.feature [hunk 1]
```diff
@@ -0,0 +1,36 @@
+@raffle
+Feature:
+  In order to give away cool prizes to our members that leave feedback
+  As a organizer
```
Comment (HcsOmot) [inline features/picking-events-to-raffle.feature]: Typo:
As an organizer

File: features/picking-events-to-raffle.feature [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      | id  | title    | eventId | importedAt          |
+      | 101 | Talk 101 | 1       | 2020-01-02 11:22:33 |
+      | 201 | Talk 201 | 2       | 2020-01-02 11:22:33 |
+      | 202 | Talk 202 | 2       | 2020-01-02 11:22:33 |
+      | 301 | Talk 301 | 3       | 2020-01-02 11:22:33 |
+    And we have these users in the system
+      | id | username | displayName |
+      | 1  | User1    | User 1      |
+      | 2  | User2    | User 2      |
+      | 3  | User3    | User 2      |
```
Comment (HcsOmot) [inline features/picking-events-to-raffle.feature]: Typo:

User3, with ID 3, should probably have display name User 3

File: features/raffling.feature [hunk 3]
```diff
@@ -0,0 +1,40 @@
+@raffle
+Feature:
+  In order to give away cool prizes to our members that leave feedback
+  As a organizer
```
Comment (HcsOmot) [inline features/raffling.feature]: As an organizer

File: features/raffling.feature [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      | id  | title    | eventId | importedAt          |
+      | 101 | Talk 101 | 1       | 2020-01-02 11:22:33 |
+      | 201 | Talk 201 | 2       | 2020-01-02 11:22:33 |
+      | 202 | Talk 202 | 2       | 2020-01-02 11:22:33 |
+      | 301 | Talk 301 | 3       | 2020-01-02 11:22:33 |
+    And we have these users in the system
+      | id | username | displayName |
+      | 1  | User1    | User 1      |
+      | 2  | User2    | User 2      |
+      | 3  | User3    | User 2      |
```
Comment (HcsOmot) [inline features/raffling.feature:23]: User3, ID 3, displayName User 3?

File: features/raffling.feature [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    And we have each user commenting on each talk
+    And there are no raffles in the system
+    And organizer picks to raffle meetups: "2,3"
+
+  Scenario: ...
```
Comment (HcsOmot) [inline features/raffling.feature]: Proposed scenario name:

Selecting a user with 3 comments removes 3 comments from the raffle

File: features/bootstrap/FixturesContext.php [hunk 6]
```diff
@@ -76,6 +86,65 @@ public function weHaveThisTalksInTheSystem(TableNode $table)
...
+        foreach ($this->getUserRepository()->findAll() as $user) {
+            $this->getEntityManager()->remove($user);
+        }
+        $this->getEntityManager()->flush();
+    }
+
+    /**
+     * @Given we have these users in the system
+     */
+    public function weHaveThisUsersInTheSystem(TableNode $table)
```
Comment (HcsOmot) [inline features/bootstrap/FixturesContext.php:119]: Typo:
method name should match Behat step definition

weHaveThisUsersInTheSystem > weHaveTheseUsersInTheSystem

File: features/picking-events-to-raffle.feature [hunk 7]
```diff
@@ -0,0 +1,36 @@
+@raffle @api
+Feature:
+  In order to give away cool prizes to our members that leave feedback
+  As a organizer
```
Comment (HcsOmot) [inline features/picking-events-to-raffle.feature]: Typo:

As a organizer > As an organizer

File: features/raffling.feature [hunk 8]
```diff
@@ -0,0 +1,45 @@
+@raffle @api
+Feature:
+  In order to give away cool prizes to our members that leave feedback
+  As a organizer
```
Comment (HcsOmot) [inline features/raffling.feature]: Typo:

As a organizer > As an organizer

File: features/raffling.feature [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    And we have each user commenting on each talk
+    And there are no raffles in the system
+    And organizer picks to raffle meetups: "2,3"
+
+
+  Scenario: ...
+    When we pick
+    Then we should get back one of the members that left feedback
+
+  Scenario: ...
```
Comment (HcsOmot) [inline features/raffling.feature:33]: Scenario declaration proposition:
Selecting a user who submitted 3 comments removes 3 comments from the raffle

File: features/raffling.feature [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    When user "User1" wins
+    Then there should be 6 comments on the raffle
+
+  Scenario: ...
```
Comment (HcsOmot) [inline features/raffling.feature:37]: Scenario declaration proposition:
Selecting a user who are not currently present removes their comments from the current raffle

**Comment by msvrtan** [comment]
Typos fixed! :man_facepalming:  Thanks for noticing them!

Will merge this in

File: src/Entity/Raffle.php [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    public function getCommentsEligibleForRaffling(): Collection
+    {
+        $comments = new ArrayCollection();
+
+        foreach ($this->events as $event) {
+            foreach ($event->getTalks() as $talk) {
+                foreach ($talk->getComments() as $comment) {
+                    if (false === $this->userAlreadyWonOrIsNoShow($comment)) {
```
Comment (HcsOmot) [inline src/Entity/Raffle.php:88]: maybe we could refactor this?
```
$this->userAlreadyWonOrIsNoShow($comment->getUser)
```

and change the method signature for userAlreadyWonOrIsNoShow() to
```
userAlreadyWonOrIsNoShow(JoindInUser $user)
```

File: src/Entity/Raffle.php [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+            'createdAt' => $this->createdAt->format('c'),
+            'events'    => $events,
+        ];
+    }
+
+    /**
+     * Checks if that commenter is already a winner or a no show so we can easily filter out which comments are
+     * eligible for raffling.
+     */
+    private function userAlreadyWonOrIsNoShow(JoindInComment $comment): bool
```
Comment (HcsOmot) [inline src/Entity/Raffle.php:142]: if we change the method signature, the implementation will need to be refactored as well

**Review by HcsOmot — COMMENTED** [review]
Something went wrong while submitting review notes.

File: features/bootstrap/FixturesContext.php (hunk 6 above)
Comment (msvrtan) [reply features/bootstrap/FixturesContext.php:119]: Ooops, will fix!

File: features/raffling.feature (hunk 9 above)
Comment (msvrtan) [reply features/raffling.feature:33]: Can you open an issue and fix it in a pr?

File: features/raffling.feature (hunk 10 above)
Comment (msvrtan) [reply features/raffling.feature:37]: Can you open an issue and fix it in a pr?

File: src/Entity/Raffle.php (hunk 11 above)
Comment (msvrtan) [reply src/Entity/Raffle.php:88]: Great idea, can you open an issue and create a PR?

File: features/raffling.feature (hunk 9 above)
Comment (HcsOmot) [reply features/raffling.feature:33]: sure.

**Pull request merged**